### DISCUSSION TITLE
Which kind of sensors "could" become supported

### DISCUSSION THREAD

**Discussion by bartbutenaers** [phoddie/node-red-mcu#51] [General]
Hi @phoddie,

I am a Node-RED contributor, and this project is kind of mindblowing.  Very cool stuff!!

Being able to replace EspEasy on my mcu's by node-red-mcu in the near future (and running Node-RED on all my devices) would be awesome.  I assume node-red-mcu doesn't support yet what I need, but just want to get a feeling whether it 'could' be possible in the future.   It is not very clear to me how it works under the cover, so I would appreciate a lot if you could give some (noob level ;-) ) background information to my questions below.

Last week we had a nice discussion on Discourse about monitoring energy via Node-RED.  Got a lot of useful feedback from the community.  The [result](https://discourse.nodered.org/t/3-phase-power-consumption-measurement/68489/130?u=bartbutenaers) was a setup with a series of ***PZEM-004t*** sensors, that are connected to one of my [WT32-ETH01](https://files.seeedstudio.com/products/102991455/WT32-ETH01_datasheet_V1.1-%20en.pdf) (which I try to use as much as possible,  because I like to have things wired).

1. I don't see the WT32-ETH01 in the list of models supported by Moddable, and of course you cannot support all kind of boards.  But do you think it will never be possible to run node-red-mcu on such devices, e.g. due to some technical limitations (regarding ethernet libraries or something alike)?

2. The PZEM-004t is also not on your list of supported sensors.  Do you think sensors like these could ever become accessible via node-red-mcu?  Or will this never be possible due to some limitations e.g. with interfacing such sensors?  

3. I did some experiments recently both with [node-red-node-pisrf](https://flows.nodered.org/node/node-red-node-pisrf) and also with EspEasy to read ultrasonic distance sensors like [HC-SR04](https://www.electroschematics.com/hc-sr04-datasheet/) and the waterproof alternative [JSN-SR04T-2.0](https://www.jahankitshop.com/getattach.aspx?id=4635&Type=Product).  Again the question if sensors like this could become supported one day in the future?

Hopefully I don't give you the impression of being greedy.  Just wondering which kind of stuff could be possible from a technical point of view.  Not sure which kind of stuff (like Js or C or ... libraries and so on) is required to support new sensors.  Of course I completely understand that free time might be one of your biggest limitations...

Thanks for reading!!
Bart

**Comment by ralphwetzel** [comment]
Hello Bart,
I did a quick check of the devices you named in your text - and can give you an indication. As @phoddie has far more experience regarding these topics, I'm sure he'll add his perspective as well:

- WT32-ETH01: Moddable SDK features full Ethernet support. The rest of the board looks quite standard. Should be feasible...
- HC-SR04 looks like a nice sensor, operated by simple `Serial` protocol. Should be feasible as well...
- PZEM-004t is operated by ModBUS, as well a serial protocol. The effort to implement this looks a bit larger, but as well doable.

Thus there's a valid chance to implement all of this via `node-red-mcu`. The probability of timely availability depends on the number of people contributing to this project ... 😁 !

**Comment by phoddie** [comment]
@bartbutenaers, welcome!

> I am a Node-RED contributor, and this project is kind of mindblowing. Very cool stuff!!

Thank you! It has been an interesting adventure to learn Node-RED by trying to reimplement parts of it. ;)

I agree with @ralphwetzel's overall sentiment: what you describe should be possible. Of course, there is some work involved.

I won't try to provide detailed answers to each of your questions, because  eEach of the topics is probably its own discussion thread to cover the details. Instead, I'll provide some background and guidance.

> I don't see the WT32-ETH01 in the list of models supported by Moddable, and of course you cannot support all kind of boards. But do you think it will never be possible to run node-red-mcu on such devices, e.g. due to some technical limitations (regarding ethernet libraries or something alike)?

At a baseline level, the Moddable SDK works with most ESP32 boards out there by compiling to the generic `esp32` platform target (rather than a specific device like `esp32/moddable_two`. Of course, the generic build only knows about the ESP32 and not about any attached peripherals, such as the Ethernet peripheral here. So, a new build platform would be needed for the W32-ETH01. That looks like it would be mostly about configuring the Ethernet. The Moddable SDK does support Ethernet, as described in a [blog post](https://blog.moddable.com/blog/ethernet/). The Ethernet part is uses , the LAN8720A, is supported by the ESP-IDF so should be straightforward to get working.

> The PZEM-004t is also not on your list of supported sensors. Do you think sensors like these could ever become accessible via node-red-mcu? Or will this never be possible due to some limitations e.g. with interfacing such sensors?

If it works with Arduino, we can make it work too. ;) It looks like the PZEM-004t communicates over serial, which is well supported in the Moddable SDK thanks to our ECMA-419 I/O work. I will note that @andycarle has a standing offer: if you send him a sensor, he'll write a basic Ecma-419 Sensor driver for it. Feel free to take him up on it. ;)

> I did some experiments recently both with [node-red-node-pisrf](https://flows.nodered.org/node/node-red-node-pisrf) and also with EspEasy to read ultrasonic distance sensors like [HC-SR04](https://www.electroschematics.com/hc-sr04-datasheet/) and the waterproof alternative [JSN-SR04T-2.0](https://www.jahankitshop.com/getattach.aspx?id=4635&Type=Product). Again the question if sensors like this could become supported one day in the future?

Yes, mostly the same answer as the PZEM-004t above.

> Not sure which kind of stuff (like Js or C or ... libraries and so on) is required to support new sensors.

All the sensor work should be in JavaScript. While you have the option to use C, it is seldom necessary. All of our ECMA-419 Sensor implementations are 100% JavaScript. You can take a look at a suite of them in [$MODDABLE/modules/drivers/sensors](https://github.com/Moddable-OpenSource/moddable/tree/public/modules/drivers/sensors) to get a sense of how they are written, their relative complexity (generally pretty low), and the [Sensor Class pattern](https://419.ecma-international.org/#-13-sensor-class-pattern).

Seems like a great project, realistic in scope and possible to approach step-by-step . The result would be useful to many others. I'd be happy to support that. You'll almost certainly end up teaching me a few things about Node-RED along the way too.

**Comment by bartbutenaers** [comment]
@phoddie, @ralphwetzel,
Thanks for the smooth introduction!

This is VERY encouraging to get started with node-red-mcu, to say the least.  

Since most of my own free time goes to Node-RED development already, I have taken the liberty to ask on [Discourse](https://discourse.nodered.org/t/anybody-experimenting-with-node-red-mcu/71836) if some folks can test this module and share their feedback.  I really hope that some volunteers popup.  E.g. a page per sensor ([like](https://www.letscontrolit.com/wiki/index.php/HC-SR04) e.g. used by EspEasy) with a description of the sensor and an example Node-RED flow would be really helpful for people (like) me to get started.  But that is all easier said than done of course...

My sensors have not been shipped yet by Aliexpress, so I have some time to think about how I will get started with this ;-)

If @andycarle should have some time/interest to have a look at implementing the above sensors, that would be very welcome!  Not because I am to lazy to contribute to this project, but at the moment this ESP32/ESP8266 related development stuff is rather far away from my comfort zone.  Some extra implementations from various sensor types might help to gain some insights.   

Is there some easy way I can pay Andy to buy such sensors?  I assume that might be more convenient/faster instead of me buying the sensor and sending it to him?

About the statement *"the LAN8720A, is supported by the ESP-IDF so should be straightforward to get working"*.   Could you explain a bit what "straightforward" means?  I have the impression that it won't be so straightforward for me ;-)

**Reply by andycarle** [reply]
I went ahead and ordered the two sensors—I'll give them a look when they get here next week and let you know how it goes.

**Reply by bartbutenaers** [reply]
That is very kind of you.  Please let me know if you need anything from my side!  Looking forward to see the results of your experiment...

**Reply by bartbutenaers** [reply]
@andycarle,
We had a long discussion about the pzem-004t sensor in the Node-RED community.  The [summary](https://discourse.nodered.org/t/3-phase-power-consumption-measurement/68489/186?u=bartbutenaers) is that N of these sensors could be wired together and read by a single ESP like this:

![image](https://user-images.githubusercontent.com/14224149/208316075-6c824a96-f4e0-44d5-987f-10b98b832aea.png)

That would be a nice setup to monitor N electrical circuits in an electricity closet.
However @Paul-Reed was wondering whether there could be collisions of data of the different sensors.

Not sure at this moment how something like that would map to node-red-mcu:
+ Is this something I need to take into account in my Node-RED flow, i.e. that I have to add N nodes to my flow and call them one by one (to avoid collisions).  
+ Or is that something that is only possible if your Sensor implementation supports something like that?

**Comment by andycarle** [comment]
I started working on these sensors a bit this week, leading off with the HC-SR04. I have that working now and will plan to get the relevant pieces moved over to the Moddable SDK early next week.

The HC-SR04 was an interesting addition to our lineup, as it needs an input type that was not previously covered by either the Moddable SDK or ECMA-419: pulse width measurement. I've put together an ECMA-419 IO implementation for pulse width on ESP32, a basic HC-SR04 driver that conforms to the ECMA-419 `Proximity` sensor class, and a trivial example that uses it. 

Next on my desk is the PZEM-004t. Once I have that going I'll also contemplate options for ganging them up, as discussed above.

**Reply by andycarle** [reply]
@bartbutenaers The above-mentioned HC-SR04 driver is now live in the Moddable SDK. [Driver](https://github.com/Moddable-OpenSource/moddable/tree/public/modules/drivers/sensors/hc-sr04) and [Example](https://github.com/Moddable-OpenSource/moddable/tree/public/examples/drivers/sensors/hc-sr04).

The PulseWidth IO type is currently only set up to work with the `esp32/moddable_two_io` platform. I could help you get it working with an alternate Moddable SDK target or, I suspect @phoddie could help get it working with node-red-mcu.

**Reply by andycarle** [reply]
@bartbutenaers I forgot to report back here, but the PZEM-004t driver is now available in the Moddable SDK. You can see a video of my example in action about 7 minutes into the [Moddable Community Call for February](https://www.youtube.com/watch?v=Utfy0LNEQn4).

**Comment by bartbutenaers** [comment]
@andycarle,
Sorry for the delay due to the christmas madness...

That example code of yours is really nice. 
If there is one Node-RED user that wouldn't like it, then I will eat my shoes for breakfast ;-)

I will create a new discussion about my installation issues, to keep things here focussed on sensors, to make sure future readers are not getting confused.  And meanwhile @ralphwetzel is helping me with some of my noob questions on the Node-RED [forum](https://discourse.nodered.org/t/node-red-mcu-plugin-v1-0-integrating-node-red-mcu-edition-into-the-node-red-editor/72693/4).  As soon as I have time to have something running, I will report back here.
